File: median_of_2_sorted.py
# // Time Complexity : O(log())
# // Space Complexity : O(1)min(m,n)
# // Did this code successfully run on Leetcode : Yes
# // Any problem you faced while coding this : Nope


# // Your code here along with comments explaining your approach

class Solution:
    def findMedianSortedArrays(self, nums1: List[int], nums2: List[int]) -> float:
        A,B = nums1,nums2
        
        total=len(nums1)+len(nums2)
        half=total//2
        
        if len(B)< len(A):
            A,B=B,A
            
        l,r=0,len(A)-1
        
        while True:
            i=(l+r)//2
            j=half-i-2
            
            Aleft=A[i] if i>=0 else float("-infinity")
            Aright=A[i+1] if (i+1) < len(A) else float("infinity")
            Bleft=B[j] if j>=0 else float("-infinity")
            Bright=B[j+1] if (j+1) < len(B) else float("infinity")
            
            # If partition is correct
            if Aleft <=Bright and Bleft<=Aright:
                
                #if odd
                if total%2:
                    return min(Aright,Bright)
                
                #even
                return (max(Aleft,Bleft) + min(Aright,Bright))/2
            
            elif Aleft>Bright:
                r=i-1
            
            else:
                l=i+1
                

File: MedianOf2Sorted.java
// Time Complexity : O(log m)
// Space Complexity : O(1)
// Did this code successfully run on Leetcode :
// Any problem you faced while coding this :

class Solution {
    public double findMedianSortedArrays(int[] nums1, int[] nums2) {
        if((nums1 == null && nums2 == null) || (nums1.length == 0 && nums2.length == 0)) return 0.0;

        int m = nums1.length;
        int n = nums2.length;

        if(n<m) {
            return findMedianSortedArrays(nums2, nums1);
        }

        int l = 0;
        int h = m;

        while(l<=h) {
            int partX = l+(h-l)/2;

            int partY = (m+n)/2 - partX;

            double l1 = partX == 0 ? Integer.MIN_VALUE : nums1[partX-1];   //if index is 1 we need to provide 0th element
            double l2 = partY == 0 ? Integer.MIN_VALUE : nums2[partY-1];
            double r1 = partX == m ? Integer.MAX_VALUE : nums1[partX];
            double r2 = partY == n ? Integer.MAX_VALUE : nums2[partY];

            if(l1 <= r2 && l2 <=r1) {
                if((m+n)%2 == 0) {
                    return (Math.max(l1, l2) + Math.min(r1, r2))/2;
                }
                return Math.min(r1, r2);
            }
            else if(l2 > r1) {
                l = partX+1;
            } else {
                h = partX-1;
            }
        }
        return 0.0;
    }
}